### PR TITLE
Cts obey max fanout

### PR DESCRIPTION
**Pull request by arthurjolo**
As reported on [#5306](https://github.com/The-OpenROAD-Project/OpenROAD/issues/5306), CTS is not honoring the fanout limit set. Even with the sinks cluster and max sink per leaf buffer also being limited.

This was happening because after CTS builds the HTree it does some optimizations on the sinks distribution on the leaf buffers, shifting some sinks from one buffer to another violating the fanout limit. This PR adds a flag to set a fanout limit on CTS, making sure that it will respect this limit.

This flag must be added on the designs that have a fanout limit set so CTS can undestand this limit.

### CODE REVIEW

**Comment by arthurjolo** [comment]
This PR will only created the flag `-max_fanout,` to fix the issue [#5306](https://github.com/The-OpenROAD-Project/OpenROAD/issues/5306) the flag must be added the ORFS designs that will be done on a separate PR.

File: src/cts/src/TritonCTS.tcl [hunk 1]
```diff
@@ -80,6 +80,7 @@ sta::define_cmd_args "clock_tree_synthesis" {[-wire_unit unit]
                                              [-sink_clustering_size] \
                                              [-sink_clustering_max_diameter] \
                                              [-sink_clustering_enable] \
+                                             [-max_fanout] \
```
Comment (precisionmoon) [inline src/cts/src/TritonCTS.tcl:83]: We want to honor SDC set_max_fanout constraint by default.   Let's add -ignore_max_fanout If user wants to ignore max fanout limit,

File: src/cts/src/HTreeBuilder.cpp [hunk 2]
```diff
@@ -1209,6 +1209,14 @@ void HTreeBuilder::run()
   // clang-format on
 }
 
+bool HTreeBuilder::isNumberOfSinksTooSmall(unsigned numSinksPerSubRegion) const
```
Comment (precisionmoon) [inline src/cts/src/HTreeBuilder.cpp:1212]: Shouldn't this API return if either of the conditions is met?   < numMaxLeafSinks_  || < options_->getMaxFanout()?

File: src/cts/src/TritonCTS.tcl (hunk 1 above)
Comment (QuantamHD) [reply src/cts/src/TritonCTS.tcl:83]: That seems ultra confusing. If a user sets this flag and the SDC flag it seems like we could offer a warning, but principle of least surprise seems to be we should honor the more specific constraint in the CTS call.

Comment (maliberty) [reply src/cts/src/TritonCTS.tcl:83]: We discussed today not needing the flag.  Please wait for the next update.

**Comment by arthurjolo** [comment]
I used the key option because I couldn't find a way for CTS to check if it should respect a max fanout limit. Looking deeper I found a sdc API that I can get the information needed, I am already working on removing the key.

File: src/cts/src/TritonCTS.cpp [hunk 3]
```diff
@@ -109,14 +109,72 @@ void TritonCTS::addBuilder(TreeBuilder* builder)
...
+    auto io_type = mterm->getIoType();
+    if (io_type == odb::dbIoType::OUTPUT) {
+      buffer_port = network_->dbToSta(mterm);
+      break;
+    }
+  }
+  if (buffer_port == nullptr) {
+    return 0;
+  }
+  openSta_->sdc()->fanoutLimit(
```
Comment (precisionmoon) [inline src/cts/src/TritonCTS.cpp:134]: Just a minor suggestion.  You can also start with a large fanout limit like INF. 
Then you can start taking, sequentially one by one, min (fanout_limit, new_fanout_limit) for port, cell, liberty port, etc.
fanout_limit = INF;
if (port has fanout limit) {
    fanout_limit = MIN (fanout_limit, new fanout limit);
}
If (cell has fanout limt) {
   fanout_limit = MIN(fanout_limit, new fanout limit);
}
If (lib port has fanout limit) {
   fanout_limit = MIN(fanout_limit, new fanout limit);
}

**Comment by arthurjolo** [comment]
The secure-CI is passing for this PR. There is only 1 design that still is having 1 max fanout violation: ihp-sg13g2/riscvi,  this violation is related to the tech characterization step, where CTS is deciding to not use buffers on some levels of the clock tree. Talking with @precisionmoon we believe it is best to merge this PR to fix the max fanout violations related with the sinks and look at the characterization as a separate issue.

**Comment by arthurjolo** [comment]
> The secure-CI is passing for this PR. There is only 1 design that still is having 1 max fanout violation: ihp-sg13g2/riscvi, this violation is related to the tech characterization step, where CTS is deciding to not use buffers on some levels of the clock tree. Talking with @precisionmoon we believe it is best to merge this PR to fix the max fanout violations related with the sinks and look at the characterization as a separate issue.

Using `-distance_between_buffers 60` on ihp-sg13g2/riscvi it was possible to solve this problem.

**Comment by maliberty** [comment]
Please make a companion ORFS PR with the necessary change so they can be merged together.

**Comment by arthurjolo** [comment]
Ok, I am going to create the PR.

File: src/cts/src/HTreeBuilder.cpp [hunk 4]
```diff
@@ -1209,6 +1210,15 @@ void HTreeBuilder::run()
...
+{
+  if (options_->getMaxFanout()) {
+    logger_->report("Ta verificando por fanout");
```
Comment (maliberty) [inline src/cts/src/HTreeBuilder.cpp:1216]: rm - Look like debug code

File: src/cts/src/TritonCTS.cpp [hunk 5]
```diff
@@ -1493,6 +1554,19 @@ odb::dbITerm* TritonCTS::getFirstInput(odb::dbInst* inst) const
...
+  odb::dbSet<odb::dbITerm> iterms = inst->getITerms();
+  for (odb::dbITerm* iterm : iterms) {
+    if (iterm->isInputSignal()) {
+      return iterm;
+    }
+  }
+
+  return nullptr;
+}
+*/
```
Comment (maliberty) [inline src/cts/src/TritonCTS.cpp:1568]: rm

**Review by maliberty — CHANGES_REQUESTED** [review]

**Review by maliberty — APPROVED** [review]

**Pull request merged**